<a href="https://colab.research.google.com/github/ojasnadkar96/cs273p_project/blob/master/multiLayerPerceptron1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Multi-layer Perceptron (168 Features)

Importing all the required libraries.<br>

In [2]:
import pickle
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
%matplotlib inline

import numpy
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier

Using TensorFlow backend.


The two functions below are to save and import pickle files.<br>

In [0]:
def save_pkl(df,name):
    fullname = name+'.pkl'
    output = open(fullname, 'wb')
    pickle.dump(df, output)
    output.close()

In [0]:
def import_pkl(df,name):
    fullname = name+'.pkl'
    df = pickle.load(open(fullname, 'rb'))
    return df

In [0]:
df_train = pd.DataFrame()
df_valid = pd.DataFrame()
df_test = pd.DataFrame()
df_train_l = pd.DataFrame()
df_valid_l = pd.DataFrame()
df_test_l = pd.DataFrame()

In [0]:
df_train = import_pkl(df_train,'train_x')
df_valid = import_pkl(df_valid,'valid_x')
df_test = import_pkl(df_test,'test_x')
df_train_l = import_pkl(df_train_l,'train_x_l')
df_valid_l = import_pkl(df_valid_l,'valid_x_l')
df_test_l = import_pkl(df_test_l,'test_x_l')

In [7]:
print(df_train.shape)
print(df_valid.shape)
print(df_test.shape)
print(df_train_l.shape)
print(df_valid_l.shape)
print(df_test_l.shape)

(77854, 168)
(13737, 168)
(10175, 168)
(77854, 1)
(13737, 1)
(10175, 1)


Pre-processed data with a total of 168 features has been imported into dataframes.

The input is set to 168 as there are 168 features in the dataset.<br>
We initialize a random model with 3 hidden layers of size 256, 512 and 256 respectively.<br>
The epochs are set to 25 and the batch size is set to 100.<br>
We will test the accuracy on train and validation for this model.<br>

In [8]:
test_model = Sequential()
test_model.add(Dense(168, input_dim=168, activation='relu'))
test_model.add(Dense(256, activation='relu'))
test_model.add(Dense(512, activation='relu'))
test_model.add(Dense(256, activation='relu'))
test_model.add(Dense(1, activation='sigmoid'))
test_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
test_model.fit(df_train, df_train_l, epochs=25, batch_size=100)

W0613 22:19:32.034279 140150942447488 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

W0613 22:19:32.056976 140150942447488 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W0613 22:19:32.060847 140150942447488 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.

W0613 22:19:32.117051 140150942447488 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/optimizers.py:790: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

W0613 22:19:32.137677 140150942447488 deprecation_wrappe

Epoch 1/25
77854/77854 [==============================] - 9s 110us/step - loss: 0.5483 - acc: 0.5112
Epoch 2/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.4993 - acc: 0.5268
Epoch 3/25
77854/77854 [==============================] - 8s 102us/step - loss: 0.4821 - acc: 0.5336
Epoch 4/25
77854/77854 [==============================] - 7s 96us/step - loss: 0.4698 - acc: 0.5330
Epoch 5/25
77854/77854 [==============================] - 8s 99us/step - loss: 0.4477 - acc: 0.5379
Epoch 6/25
77854/77854 [==============================] - 8s 100us/step - loss: 0.4420 - acc: 0.5395
Epoch 7/25
77854/77854 [==============================] - 8s 100us/step - loss: 0.4138 - acc: 0.5435
Epoch 8/25
77854/77854 [==============================] - 7s 94us/step - loss: 0.4149 - acc: 0.5422
Epoch 9/25
77854/77854 [==============================] - 8s 97us/step - loss: 0.3985 - acc: 0.5443
Epoch 10/25
77854/77854 [==============================] - 7s 94us/step - loss: 0.3877 - acc: 0

In [9]:
score_train = test_model.evaluate(df_train,df_train_l)
print("score: ", score_train[1]*100)

77854/77854 [==============================] - 3s 40us/step
score:  55.28296555074985


In [10]:
score_valid = test_model.evaluate(df_valid,df_valid_l)
print("score: ", score_valid[1]*100)

13737/13737 [==============================] - 1s 40us/step
score:  52.194802358593584


In [11]:
score_test = test_model.evaluate(df_test,df_test_l)
print("score: ", score_test[1]*100)

10175/10175 [==============================] - 0s 38us/step
score:  52.99262899262899


The train, validation and test accuracy is: 55.2%, 52.1% and 52.9% respectively.<br>
Let us see if we can increase this accuracy by tuning the parameters.<br>

### Optimizer

We will tune the optimizer.<br>
We have options between `adam` and `SGD`<br>
We will use GridSearchCV on top of Keras to tune the parameter.<br>
The epochs and batchsize is kept the same as for above model.<br>

In [0]:
def nn_model_1(optimizer='adam'):
  model_1 = Sequential()
  model_1.add(Dense(168, input_dim=168, activation='relu'))
  model_1.add(Dense(256, activation='relu'))
  model_1.add(Dense(512, activation='relu'))
  model_1.add(Dense(256, activation='relu'))
  model_1.add(Dense(1, activation='sigmoid'))
  model_1.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
  return model_1

In [0]:
model_1 = KerasClassifier(build_fn=nn_model_1, epochs=25, batch_size=100, verbose=1)

In [0]:
optimizer = ['SGD', 'Adam']
param_grid_1 = dict(optimizer=optimizer)

In [0]:
grid_2 = GridSearchCV(estimator=model_1, param_grid=param_grid_1, n_jobs=4, verbose=1, cv=2)
grid_2.fit(df_train,df_train_l)
grid_2.best_params_

Fitting 2 folds for each of 2 candidates, totalling 4 fits


[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done   2 out of   4 | elapsed:  5.1min remaining:  5.1min
[Parallel(n_jobs=4)]: Done   4 out of   4 | elapsed:  5.6min finished


Epoch 1/25
77854/77854 [==============================] - 10s 122us/step - loss: 0.6513 - acc: 0.4213
Epoch 2/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.5780 - acc: 0.5126
Epoch 3/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.5374 - acc: 0.5350
Epoch 4/25
77854/77854 [==============================] - 8s 106us/step - loss: 0.5174 - acc: 0.5385
Epoch 5/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.5048 - acc: 0.5425
Epoch 6/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.4942 - acc: 0.5429
Epoch 7/25
77854/77854 [==============================] - 8s 103us/step - loss: 0.4855 - acc: 0.5414
Epoch 8/25
77854/77854 [==============================] - 8s 102us/step - loss: 0.4775 - acc: 0.5441
Epoch 9/25
77854/77854 [==============================] - 8s 100us/step - loss: 0.4707 - acc: 0.5447
Epoch 10/25
77854/77854 [==============================] - 8s 101us/step - loss: 0.4628 - 

{'optimizer': 'SGD'}

Optimal optmizer is `SGD`<br>
We'll use this from now on.<br>

### Initialization

We will tune the Initialization.<br>
We have options between `uniform` and `normal`<br>
We will use GridSearchCV on top of Keras to tune the parameter.<br>
The epochs and batchsize is kept the same as for above model.<br>

In [0]:
def nn_model_2(init_mode='uniform'):
  model_2 = Sequential()
  model_2.add(Dense(168, input_dim=168, kernel_initializer=init_mode, activation='relu'))
  model_2.add(Dense(256, kernel_initializer=init_mode, activation='relu'))
  model_2.add(Dense(512, kernel_initializer=init_mode, activation='relu'))
  model_2.add(Dense(256, kernel_initializer=init_mode, activation='relu'))
  model_2.add(Dense(1, kernel_initializer=init_mode, activation='sigmoid'))
  model_2.compile(loss='binary_crossentropy', optimizer='SGD', metrics=['accuracy'])
  return model_2

In [0]:
model_2 = KerasClassifier(build_fn=nn_model_2, epochs=25, batch_size=100, verbose=1)

In [0]:
init_mode = ['uniform','normal']
param_grid_2 = dict(init_mode=init_mode)

In [0]:
grid_2 = GridSearchCV(estimator=model_2, param_grid=param_grid_2, n_jobs=4, verbose=1, cv=2)
grid_2.fit(df_train,df_train_l)
grid_2.best_params_

Fitting 2 folds for each of 2 candidates, totalling 4 fits


[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done   2 out of   4 | elapsed:  4.9min remaining:  4.9min
[Parallel(n_jobs=4)]: Done   4 out of   4 | elapsed:  5.0min finished


Epoch 1/25
77854/77854 [==============================] - 10s 127us/step - loss: 0.6850 - acc: 0.3493
Epoch 2/25
77854/77854 [==============================] - 8s 107us/step - loss: 0.6824 - acc: 0.3493
Epoch 3/25
77854/77854 [==============================] - 8s 107us/step - loss: 0.6821 - acc: 0.3493
Epoch 4/25
77854/77854 [==============================] - 8s 105us/step - loss: 0.6816 - acc: 0.3493
Epoch 5/25
77854/77854 [==============================] - 8s 102us/step - loss: 0.6808 - acc: 0.3493
Epoch 6/25
77854/77854 [==============================] - 8s 102us/step - loss: 0.6792 - acc: 0.3493
Epoch 7/25
77854/77854 [==============================] - 8s 101us/step - loss: 0.6750 - acc: 0.3493
Epoch 8/25
77854/77854 [==============================] - 8s 102us/step - loss: 0.6599 - acc: 0.3541
Epoch 9/25
77854/77854 [==============================] - 8s 103us/step - loss: 0.6108 - acc: 0.4776
Epoch 10/25
77854/77854 [==============================] - 8s 105us/step - loss: 0.5566 - 

{'init_mode': 'uniform'}

Optimal Initialization is `uniform`<br>
We'll use this from now on.<br>

### Activation Function

We will tune the Activation Function.<br>
We have options between `relu` and `softmax`<br>
We will use GridSearchCV on top of Keras to tune the parameter.<br>
The epochs and batchsize is kept the same as for above model.<br>

In [0]:
def nn_model_3(activation='relu'):
  model_3 = Sequential()
  model_3.add(Dense(168, input_dim=168, kernel_initializer='uniform', activation=activation))
  model_3.add(Dense(256, kernel_initializer='uniform', activation=activation))
  model_3.add(Dense(512, kernel_initializer='uniform', activation=activation))
  model_3.add(Dense(256, kernel_initializer='uniform', activation=activation))
  model_3.add(Dense(1, kernel_initializer='uniform', activation='sigmoid'))
  model_3.compile(loss='binary_crossentropy', optimizer='SGD', metrics=['accuracy'])
  return model_3

In [0]:
model_3 = KerasClassifier(build_fn=nn_model_3, epochs=25, batch_size=100, verbose=1)

In [0]:
activation = ['softmax','relu']
param_grid_3 = dict(activation=activation)

In [0]:
grid_3 = GridSearchCV(estimator=model_3, param_grid=param_grid_3, n_jobs=4, verbose=1, cv=2)
grid_3.fit(df_train,df_train_l)
grid_3.best_params_

Fitting 2 folds for each of 2 candidates, totalling 4 fits


[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done   2 out of   4 | elapsed:  4.9min remaining:  4.9min
[Parallel(n_jobs=4)]: Done   4 out of   4 | elapsed:  5.2min finished


Epoch 1/25
77854/77854 [==============================] - 10s 124us/step - loss: 0.6851 - acc: 0.3495
Epoch 2/25
77854/77854 [==============================] - 8s 102us/step - loss: 0.6825 - acc: 0.3493
Epoch 3/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.6822 - acc: 0.3493
Epoch 4/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.6819 - acc: 0.3493
Epoch 5/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.6814 - acc: 0.3493
Epoch 6/25
77854/77854 [==============================] - 8s 103us/step - loss: 0.6804 - acc: 0.3493
Epoch 7/25
77854/77854 [==============================] - 8s 100us/step - loss: 0.6783 - acc: 0.3493
Epoch 8/25
77854/77854 [==============================] - 8s 102us/step - loss: 0.6720 - acc: 0.3493
Epoch 9/25
77854/77854 [==============================] - 8s 103us/step - loss: 0.6467 - acc: 0.3766
Epoch 10/25
77854/77854 [==============================] - 8s 102us/step - loss: 0.5851 - 

{'activation': 'relu'}

Optimal Activation Function is `relu`<br>
We'll use this from now on.<br>

### Final Model

We will use the tuned parameters and check accuracy.<br>

In [0]:
model_final = Sequential()
model_final.add(Dense(168, input_dim=168, kernel_initializer='uniform', activation='relu'))
model_final.add(Dense(256, kernel_initializer='uniform', activation='relu'))
model_final.add(Dense(512, kernel_initializer='uniform', activation='relu'))
model_final.add(Dense(256, kernel_initializer='uniform', activation='relu'))
model_final.add(Dense(1, kernel_initializer='uniform', activation='sigmoid'))
model_final.compile(loss='binary_crossentropy', optimizer='SGD', metrics=['accuracy'])
model_final.fit(df_train,df_train_l,epochs=25, batch_size=100, verbose=1)

Epoch 1/25
77854/77854 [==============================] - 10s 128us/step - loss: 0.6850 - acc: 0.3496
Epoch 2/25
77854/77854 [==============================] - 8s 106us/step - loss: 0.6824 - acc: 0.3493
Epoch 3/25
77854/77854 [==============================] - 8s 106us/step - loss: 0.6820 - acc: 0.3493
Epoch 4/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.6815 - acc: 0.3493
Epoch 5/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.6807 - acc: 0.3493
Epoch 6/25
77854/77854 [==============================] - 8s 104us/step - loss: 0.6789 - acc: 0.3493
Epoch 7/25
77854/77854 [==============================] - 8s 105us/step - loss: 0.6738 - acc: 0.3493
Epoch 8/25
77854/77854 [==============================] - 8s 107us/step - loss: 0.6547 - acc: 0.3651
Epoch 9/25
77854/77854 [==============================] - 8s 105us/step - loss: 0.5983 - acc: 0.4950
Epoch 10/25
77854/77854 [==============================] - 8s 106us/step - loss: 0.5468 - 

In [0]:
score_train_f = model_final.evaluate(df_train,df_train_l)
print("score: ", score_train_f[1]*100)

77854/77854 [==============================] - 6s 80us/step
score:  56.07033678500276


In [0]:
score_valid_f = model_final.evaluate(df_valid,df_valid_l)
print("score: ", score_valid_f[1]*100)

13737/13737 [==============================] - 1s 72us/step
score:  53.068355536143265


In [0]:
score_test_f = model_final.evaluate(df_test,df_test_l)
print("score: ", score_test_f[1]*100)

10175/10175 [==============================] - 1s 74us/step
score:  53.631449631449634


The train, validation and test accuracy is: 56%, 53.1% and 53.6% respectively.<br>
This seems like the better model and therfore we keep these hyperparameters.<br>